### ISSUE TITLE
Transpile bindings for `u32` map to `number` in *.d.ts

### ISSUE THREAD

**Issue by nuke-web3** [bytecodealliance/jco#402]
I noticed in making [this little demo](https://github.com/bytecodealliance/component-docs/tree/main/component-model/examples/tutorial/jco) that I could _not_ use signed or out-of-bounds numbers for `wasmtime` but that in the browser and via node it does.

https://stackoverflow.com/help/minimal-reproducible-example adding to [the cli](https://github.com/bytecodealliance/component-docs/blob/main/component-model/examples/tutorial/jco/cli-calc.js):

`../wit/calculator.wit`
```wit
interface calculate {
    enum op {
        add,
    }
    eval-expression: func(op: op, x: u32, y: u32) -> u32;
}
```

`jco` version 1.0.3 generated `bindings/interfaces/docs-calculator-calculate.d.ts`
```js
export namespace DocsCalculatorCalculate {
  export function evalExpression(op: Op, x: number, y: number): number;
}
```

`cli-calc.js:`
```js
// u32 is specified in wit, but typescript bindings specify only `number`
console.log("Unsigned overflow = " + calculate.evalExpression("add", 4294967300, 0));
console.log("Unsigned underflow = " + calculate.evalExpression("add", -4294967200, 0));
//                                   negative should fail types... 😬 ^^^^^
```
```
// $ node cli-calc.js
Answer (to life) = 42
Unsigned overflow = 4
Unsigned underflow = 96
```

Is there a way now to ensure that bindings generated use the WIT specified  stronger types? I would think softening the interface to `number` should not be the default :sweat_smile:

**Comment by guybedford** [comment]
U32 lowering does coercion currently - https://github.com/bytecodealliance/jco/blob/main/crates/js-component-bindgen/src/function_bindgen.rs#L288.

I agree though that we should move to validation with traps for all user-provided values.

**Comment by NukeManDan** [comment]
> ```js
> // Use `>>>0` to ensure the bits of the number are treated as unsigned.
> Instruction::U32FromI32 => results.push(format!("{} >>> 0", operands[0])),

Curious, IIUC the intent here is to ensure no negative numbers should be possible, yet in my example above,at runtime we happily underflow with a negative number passed in... Should that trap? 

I would be a strong advocate for the behavior to try and do its best to fail at runtime if wit import / export type properties are violated (by default at least). 

As it is now, devs need to make those assertions wrapping the bindings in JS, correct?

**Comment by guybedford** [comment]
Yes, currently the trap behaviour is inconsistent, some type errors trap and others coerce.

**Comment by jocrau** [comment]
This inconsistency lead to a somewhat unexpected behavior in my code. I totally agree with @nuke-web3 when they state "I would be a strong advocate for the behavior to try and do its best to fail at runtime if wit import / export type properties are violated (by default at least)." Shouldn't the behavior be at least as strict as the generated TypeScript types require it? Is there a plan to change this.

**Comment by vados-cosmonic** [comment]
Note: [jco now has a `--strict` option](https://github.com/bytecodealliance/jco/pull/1403) (which will be made the default in the next major version), *some* checks have been added where we used to simply coerce (`toUint32` and friends -- see `conversion.rs`), bu the work is as of yet incomplete (e.g. `Instruction::U32FromI32` does not yet use the `--strict` setting and fail loudly.

**Comment by vados-cosmonic** [comment]
OK, now that we have `--strict` mode, we can give out better errors here. It's not the default yet but will be at the next breaking release.

We also have a regression test being being added:
https://github.com/bytecodealliance/jco/pull/1923
